### PR TITLE
BRS-288: Add SMS Messaging

### PR DESCRIPTION
**Pull request by Christopher-walsh22**
### Jira Ticket:

BRS-288

### Jira Ticket URL:

[BRS-288: Opt in to SMS reminder in DUP Public](https://github.com/orgs/bcgov/projects/49/views/1?pane=issue&itemId=43920615)

### Description:
- Update form to include checkbox to opt in to SMS reminder
- Update form to include field for phone number
- Logic to handle phone number on submission (include if opted in / validate if opted in / remove hyphens upon submit)
- Include new directive to format phone numbers and validate information provided
- Tests to ensure phone number length is limited and clear phone number input upon opting out
- Update regex validation on phone number field to include hyphens

### CODE REVIEW

File: src/app/registration/contact-form/contact-form.component.html [hunk 1]
```diff
@@ -79,7 +83,35 @@ <h3>
...
+            name="phone" 
+            placeholder="111-222-3333"
+            maxlength="12"
+            appPhoneFormat 
+            />
+          <div *ngIf="myForm.errors?.['phoneRequired']">
+            <small *ngIf="myForm.errors?.['phoneRequired']" class="text-danger">*Your Phone Number is Required</small>
+          </div>
+          <div *ngIf="((phone.invalid) && phone.dirty) && enablePhone">
+            <small *ngIf="phone.errors?.pattern" class="text-danger">*You have selected to be notifyied by SMS. You must enter a valid phone number.</small>
```
Comment (cameronpettit) [inline src/app/registration/contact-form/contact-form.component.html:112]: nit: `notified`

File: src/app/shared/components/phone-number-formater/phoneFormat.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@Directive({
+  selector: '[appPhoneFormat]'
+})
+export class PhoneFormatDirective {
+
+  constructor(private el: ElementRef) {}
+
+  @HostListener('input', ['$event']) 
+  @HostListener('keydown', ['$event'])
+  onInput(event: any) {
```
Comment (cameronpettit) [inline src/app/shared/components/phone-number-formater/phoneFormat.ts:12]: This works for most cases. Another option is to subscribe to changes in the form value, so the formatter fires even if someone pastes a number into the field (no `keydown` event). The downside of that is if the control value is ever invalid then you have to take an additional step to set it to something valid again (no event to stop propagation on).

**Review by cameronpettit — APPROVED** [review]
approve w/nit

File: src/app/shared/components/phone-number-formater/phoneFormat.ts (hunk 2 above)
Comment (cameronpettit) [reply src/app/shared/components/phone-number-formater/phoneFormat.ts:12]: I wouldn't worry about changing anything here for this PR though

Comment (Christopher-walsh22) [reply src/app/shared/components/phone-number-formater/phoneFormat.ts:12]: Thanks, I will take that into consideration next time!

File: src/app/registration/contact-form/contact-form.component.html (hunk 1 above)
Comment (Christopher-walsh22) [reply src/app/registration/contact-form/contact-form.component.html]: resolved in: [ae18a22](https://github.com/bcgov/parks-reso-public/pull/294/commits/ae18a220c903f805922cd3a11244f132c379eddd)

**Pull request merged**